app/models/provider_label_tag_mapping.rb

```diff
@@ -0,0 +1,96 @@
+class ProviderLabelTagMapping < ApplicationRecord
+  belongs_to :tag
+  belongs_to :mappable_entity
+
+  # Return a flat list of mappable entity names. The nil is required by the
+  # user interface drop down menu for "all".
+  #
+  def self.mappable_entities(provider = 'kubernetes')
+    MappableEntity.where(:provider => provider).pluck(:name).sort.unshift(nil)
+  end
+
+  # Pass the data this returns to map_* methods.
+  def self.cache
+    # {[name, type, value] => [tag_id, ...]}
+    in_my_region.find_each
+                .group_by { |m| [m.label_name, m.labeled_resource_type, m.label_value].freeze }
+                .transform_values { |mappings| mappings.collect(&:tag_id) }
+  end
+
+  # We expect labels to be {:name, :value} hashes
+  # and return {:tag_id} or {:category_tag_id, :entry_name, :entry_description} hashes.
+
+  def self.map_labels(cache, type, labels)
+    labels.collect_concat { |label| map_label(cache, type, label) }.uniq
+  end
+
+  def self.map_label(cache, type, label)
+    # Apply both specific-type and any-type, independently.
+    (map_name_type_value(cache, label[:name], type, label[:value]) +
+     map_name_type_value(cache, label[:name], nil,  label[:value]))
+  end
+
+  def self.map_name_type_value(cache, name, type, value)
+    specific_value = cache[[name, type, value]] || []
+    any_value      = cache[[name, type, nil]]   || []
```

Comment:
[Same provider comment](https://github.com/ManageIQ/manageiq/pull/14236/files#r105028760) for the cache lookup.